### PR TITLE
Add import resolve diagnostics and clean up API

### PR DESCRIPTION
**Pull request by jmbockhorst**
Some major refactoring of the type checker api for better handling diagnostics with an async option and allowing us to calculate other diagnostics like import resolves.

### CODE REVIEW

File: src/cancellation.ts [hunk 1]
```diff
@@ -187,6 +187,21 @@ export interface ICancellationToken {
...
+export class ServerCancellationToken implements ICancellationToken {
+  constructor(private cancellationToken: CancellationToken) {}
+
+  public isCancellationRequested(): boolean {
+    return this.cancellationToken.isCancellationRequested;
+  }
+
+  public throwIfCancellationRequested(): void {
+    if (this.isCancellationRequested()) {
+      console.log("throwing cancellation");
```
Comment (Razzeee) [inline src/cancellation.ts:199]: Should this be a console log? I thought we should use the connection to log?

File: src/util/types/typeChecker.ts [hunk 2]
```diff
@@ -37,6 +38,25 @@ export interface DefinitionResult {
   nodeType: NodeType;
 }
 
+class DiagnosticsCollection extends Map<string, Diagnostic[]> {
```
Comment (Razzeee) [inline src/util/types/typeChecker.ts:41]: I guess `ElmDiagnosticsHelper` should be merged with this? (not in this PR)

File: src/cancellation.ts (hunk 1 above)
Comment (jmbockhorst) [reply src/cancellation.ts]: Meant to remove that.

File: src/util/types/typeChecker.ts (hunk 2 above)
Comment (jmbockhorst) [reply src/util/types/typeChecker.ts:41]: Yeah and we'll need to talk about the role of elm make (on save, never, controlled by a setting, etc) longer term.

**Pull request merged**